esp-hal/src/soc/esp32s3/psram.rs

```diff
@@ -168,16 +168,17 @@ pub(crate) fn init_psram(config: PsramConfig) {
         const MMU_INVALID: u32 = 1 << 14;
         const DR_REG_MMU_TABLE: u32 = 0x600C5000;
 
-        // calculate the PSRAM start address to map
-        let mut start = EXTMEM_ORIGIN;
+        // calculate the PSRAM start address to map - honor the fact there might be
+        // unmapped pages in between
         let mmu_table_ptr = DR_REG_MMU_TABLE as *const u32;
-        for i in 0..FLASH_MMU_TABLE_SIZE {
+        let mut mapped_pages = 0;
+        for i in (0..FLASH_MMU_TABLE_SIZE).rev() {
             if mmu_table_ptr.add(i).read_volatile() != MMU_INVALID {
-                start += MMU_PAGE_SIZE;
-            } else {
+                mapped_pages = i;
                 break;
             }
         }
+        let start = EXTMEM_ORIGIN + MMU_PAGE_SIZE * (mapped_pages + 1) as u32;
```

Comment:
Is the `+ 1` required here? As far as I understand, if there are no pages mapped, then just use `EXTMEM_ORIGIN` as the start value? Or maybe I just don't understand the issue well enough :D.

I think adding some brackets around `MMU_PAGE_SIZE * (mapped_pages + 1)` might help for clarity.

Reply:
Given the bootloader mapped this
```
0 MAPPED to PAGE 0 at ORIGIN + 0x00000 - 0x0ffff
1 MAPPED to PAGE 1 at ORIGIN + 0x10000 - 0x1ffff
2 MAPPED to PAGE 2 at ORIGIN + 0x20000 - 0x2ffff
```

We would end-up with  with `mapped_pages` = 2

Then we need (2 +1) * 0x10000 to get ORIGIN + 0x30000

But I agree it looks awkward - will change to make this less weird